### PR TITLE
Implement GraphARM model from [Kong et al. 2023]

### PR DESCRIPTION
**Pull request by caio-freitas**
Implement Autoregressive Diffusion Model for Graph Generation model from Kong et al.

Reopened after renaming branch.

### CODE REVIEW

File: benchmarks/GraphARM/grapharm.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+                # predictions & loss
+                for diffusion_trajectory in diffusion_trajectories:
+                    G_0 = diffusion_trajectory[0]
+                    node_order = self.node_decay_ordering(G_0) # Due to permutation invariance, we can use sample from uniform distribution
+                    for t in range(len(node_order)):
+                        for k in range(t+1):# until t
+                            node = node_order[k]
+                            G_k = diffusion_trajectory[k].clone()
+                            G_tplus1 = diffusion_trajectory[t+1].clone()
+                            G_pred = G_tplus1.clone()
```
Comment (caio-freitas) [inline benchmarks/GraphARM/grapharm.py:127]: The $G'_t$ graph used for generation contains only one masked node, connected by all other nodes by masked edges.
image

@Alonso94 @anindex one thing that's not clear to me is if they use this for training or only for sampling. ($G'_t$ does not show up again in the paper at all)

File: benchmarks/GraphARM/grapharm.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        # wandb.log({"reward": acc_reward})
+        # # update parameters (REINFORCE algorithm)
+        # self.ordering_optimizer.step()
+        
+
+
+    def step_loss(self, G_0, node_type_probs, edge_type_probs, w_k, node, node_order, t, M):
+        T = len(node_order)
+        n_i = G_0.x.shape[0]
+        edge_type_probs = torch.index_select(edge_type_probs, 0, torch.tensor(node_order[t:]).to(self.device))  # filter to only connections to previously denoised nodes
```
Comment (caio-freitas) [inline benchmarks/GraphARM/grapharm.py:198]: If $G'_t$ is used, the nodes $\geq t$ (not yet absorbed) don't have to be filtered. $G'_t$ already contains all unmasked nodes and the one to be unmasked.

File: benchmarks/GraphARM/grapharm.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        wandb.log({"p_nodes_2": node_type_probs[1].detach().cpu().numpy()})
+        wandb.log({"p_nodes_3": node_type_probs[2].detach().cpu().numpy()})
+        wandb.log({"p_nodes_4": node_type_probs[3].detach().cpu().numpy()})
+        wandb.log({"p_nodes_5": node_type_probs[4].detach().cpu().numpy()})
+        # calculate loss
+        log_p_O_v =  log_p_edges + torch.log(node_type_probs[G_0.x[node]])
+        loss = -(n_i/T)*log_p_O_v*w_k/M # cumulative, to join (k) from all previously denoised nodes
+        return loss
+
+    def step_reward(self, G_t, node_type_probs, edge_type_probs, w_k, node, node_order, t, M):
```
Comment (caio-freitas) [inline benchmarks/GraphARM/grapharm.py:218]: Both reward and loss functions can be replaced by a `VLB()` function.

File: benchmarks/GraphARM/models.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
+import torch
+from torch import nn
+from torch_geometric.nn import GAT
+from torch_geometric.utils import add_self_loops, degree
+from torch.nn import functional as F
+from torch.nn import Linear, ReLU
+import math
+from torch_geometric.nn import MessagePassing
+
+class DiffusionOrderingNetwork(nn.Module):
```
Comment (caio-freitas) [inline benchmarks/GraphARM/models.py:10]: > We use a 3-layer relational graph convolutional network with hidden dimension 256.

File: benchmarks/GraphARM/models.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        graph_embedding = graph_embedding.repeat(h_v.shape[0], 1)
+
+        node_pred = self.node_pred_layer(torch.cat([graph_embedding, h_v], dim=1)) # hidden_dim + 1
+        # aggregate with torch mean pooling
+        node_pred = torch.mean(node_pred, dim=0)
+
+        
+        # edge prediction follows a mixture of multinomial distribution, with
+        # the Softmax(sum(mlp_alpha([graph_embedding, h_vi, h_vj])))
+        alphas = F.softmax(torch.sum(self.mlp_alpha(torch.cat([graph_embedding, h_v], dim=1)), dim=0, keepdim=True), dim=1)
```
Comment (caio-freitas) [inline benchmarks/GraphARM/models.py:176]: In the paper, both $h_{v_t}$ and $h_{v_j}$ are used. Maybe this **should** be parallelized as well. 
image
Calculating the embedding $h_{v_j}$ for all nodes, $\textbf{h}$, and duplicating $h_G$ and $h_{v_t}$ to the same dimension should return $p(e_{v_t}, \textbf{v} | G_{t+1})$

File: benchmarks/GraphARM/debug.py
Comment (caio-freitas) [inline benchmarks/GraphARM/debug.py:1]: Should be deleted before merging

File: benchmarks/GraphARM/grapharm.py [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+        self.denoising_network.to(device)
+        self.masker = NodeMasking(dataset)
+
+
+
+        self.denoising_optimizer = torch.optim.Adam(self.denoising_network.parameters(), lr=1e-3, betas=(0.9, 0.999))
+        self.ordering_optimizer = torch.optim.Adam(self.diffusion_ordering_network.parameters(), lr=5e-3, betas=(0.9, 0.999))
+
+
+    def node_decay_ordering(self, datapoint):
```
Comment (caio-freitas) [inline benchmarks/GraphARM/grapharm.py:39]: Must actually sample from diffusion ordering model before merging

File: benchmarks/GraphARM/grapharm.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        self.diffusion_ordering_network = diffusion_ordering_network
+        self.diffusion_ordering_network.to(device)
+        self.denoising_network = denoising_network
+        self.denoising_network.to(device)
+        self.masker = NodeMasking(dataset)
+
+
+
+        self.denoising_optimizer = torch.optim.Adam(self.denoising_network.parameters(), lr=1e-3, betas=(0.9, 0.999))
+        self.ordering_optimizer = torch.optim.Adam(self.diffusion_ordering_network.parameters(), lr=5e-3, betas=(0.9, 0.999))
```
Comment (caio-freitas) [inline benchmarks/GraphARM/grapharm.py:36]: Is the Adam optimizer actually adequate for the REINFORCE algorithm?

**Review by caio-freitas — COMMENTED** [review]
After changing those, should be ready to merge

File: benchmarks/GraphARM/grapharm.py [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+                            G_tplus1 = diffusion_trajectory[t+1].clone()
+                            G_pred = G_tplus1.clone()
+                              
+
+                            # predict node type
+                            # logger.debug(f"Denoising node: {node}")
+                            node_type_probs, edge_type_probs = self.denoising_network(G_pred.x, G_pred.edge_index, G_pred.edge_attr)
+                            # logger.debug(f"Node type probs: {node_type_probs}")
+                            # logger.debug(f"Edge type probs: {edge_type_probs}")
+                            w_k = self.diffusion_ordering_network(G_pred)[node]
```
[the remark this answers is not in the document]
Comment (caio-freitas) [reply benchmarks/GraphARM/grapharm.py]: Outsourced for issue #51

File: benchmarks/GraphARM/models.py [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+
+        # # Positional encoding
+        for t in range(len(node_order)):
+            h[node_order[t], :] += self.pe[t, :].to(self.device)
+
+        if unmasked.numel() > 0:
+            h_not_absorbed = h[unmasked, :]
+            # softmax: h over h_not_absorbed
+            # make sure values are positive and sum to 1 (for unmasked nodes)
+            h = torch.exp(h) / torch.sum(torch.exp(h_not_absorbed), dim=0)
```
Comment (caio-freitas) [inline benchmarks/GraphARM/models.py:113]: Return only elements affected by the softmax normalization (not absorbed nodes). Avoids misuse of diffusion ordering network.

File: benchmarks/GraphARM/grapharm.py [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+                    G_0 = diffusion_trajectory[0]
+                    node_order = self.node_decay_ordering(G_0)
+                    for t in range(len(node_order)):
+                        for k in range(t+1):# until t
+                            G_pred = diffusion_trajectory[t+1].clone()                              
+
+                            # predict node and edge type distributions
+                            node_type_probs, edge_type_probs = self.denoising_network(G_pred.x, G_pred.edge_index, G_pred.edge_attr)
+
+                            w_k = self.diffusion_ordering_network(G_0, node_order[t+1:])[node_order[k]]
```
Comment (caio-freitas) [inline benchmarks/GraphARM/grapharm.py:134]: Make sure $w_k$ is detached

**Review by caio-freitas — COMMENTED** [review]
Last changes before merging

File: benchmarks/GraphARM/models.py (hunk 5 above)
Comment (caio-freitas) [reply benchmarks/GraphARM/models.py]: Issue #51
